### PR TITLE
binding: support YAML format

### PR DESCRIPTION
**Pull request by Germiniku**
### Describe the pull request

binding support yaml validator

Link to the issue: https://github.com/flamego/flamego/issues/72

### Checklist

- [x] I agree to follow this project's [Code of Conduct](https://golang.org/conduct) by submitting this pull request.
- [x] I totally understand sending a pull request has no guarantee to be merged, and the review rounds may take a long time.
- [x] I have added test cases to cover the new code.

### CODE REVIEW

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for your PR!

However, please first fill out the PR description using the [pre-filled template](https://github.com/flamego/binding/blob/main/.github/pull_request_template.md).

We can only review after that's been done.

**Comment by Germiniku** [comment]
> Thank you for submitting this PR!
> 
> Unfortunately, we are not able to review changes until you have time to take the little extra effort to fill out the PR description using the [pre-filled template](https://github.com/flamego/binding/blob/main/.github/pull_request_template.md).

I've changed it to this template now

File: binding.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ import (
 
 	"github.com/flamego/flamego"
 	"github.com/flamego/validator"
+	"gopkg.in/yaml.v3"
```
Comment (unknwon) [inline binding.go:17]:
```suggestion

	"gopkg.in/yaml.v3"

	"github.com/flamego/flamego"
	"github.com/flamego/validator"
```

ref: https://github.com/unknwon/go-code-convention/blob/main/en-US.md#import-packages

File: binding.go [hunk 2]
```diff
@@ -118,6 +119,45 @@ func JSON(model interface{}, opts ...Options) flamego.Handler {
 	})
 }
 
+// Yaml returns a middleware handler that injects a new instance of the model
+// with populated fields and binding.Errors for any deserialization, binding, or
+// validation errors into the request context. The model instance fields are
+// populated by deserializing the Yaml payload from the request body.
+func Yaml(model interface{}, opts ...Options) flamego.Handler {
```
Comment (unknwon) [inline binding.go:126]:
```suggestion
// YAML returns a middleware handler that injects a new instance of the model
// with populated fields and binding.Errors for any deserialization, binding, or
// validation errors into the request context. The model instance fields are
// populated by deserializing the YAML payload from the request body.
func YAML(model interface{}, opts ...Options) flamego.Handler {
```

Please also fix other occurrences.

File: binding.go [hunk 3]
```diff
@@ -118,6 +119,45 @@ func JSON(model interface{}, opts ...Options) flamego.Handler {
...
+					},
+				)
+			}
+		}
+		validateAndMap(c, opt.Validator, obj, errs)
+		errs = c.Value(reflect.TypeOf(errs)).Interface().(Errors)
+		if len(errs) > 0 && opt.ErrorHandler != nil {
+			_, err := c.Invoke(opt.ErrorHandler)
+			if err != nil {
+				panic("binding.Yaml: " + err.Error())
```
Comment (unknwon) [inline binding.go:155]:
```suggestion
				panic("binding.YAML: " + err.Error())
```

File: binding_test.go [hunk 4]
```diff
@@ -346,6 +346,93 @@ func TestJSON(t *testing.T) {
 	})
 }
 
+func TestYaml(t *testing.T) {
```
Comment (unknwon) [inline binding_test.go:349]:
```suggestion
func TestYAML(t *testing.T) {
```

File: binding_test.go [hunk 5]
```diff
@@ -346,6 +346,93 @@ func TestJSON(t *testing.T) {
...
+			func() {
+				type yaml struct {
+					Username string
+					Password string
+				}
+				Yaml(&yaml{})
+			},
+		)
+	})
+
```
Comment (unknwon) [inline binding_test.go:362]: Looks like a test similar to ["invalid JSON"](https://github.com/flamego/binding/blob/dbd21dca85f5a72c9ffe7fcfac551f5a40f93ac4/binding_test.go?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L36) is missing?

File: binding_test.go [hunk 6]
```diff
@@ -346,6 +346,93 @@ func TestJSON(t *testing.T) {
...
+			c.ResponseWriter().Write([]byte(fmt.Sprintf("Oops! Error occurred: %v", errs[0].Err)))
+		}
+		tests := []struct {
+			name       string
+			payload    string
+			handler    flamego.Handler
+			statusCode int
+			want       string
+		}{
+			{
```
Comment (unknwon) [inline binding_test.go:383]: Also missing a similar test to "invalid JSON"?

File: go.mod [hunk 7]
```diff
@@ -12,4 +12,5 @@ require (
 	golang.org/x/sys v0.0.0-20210917161153-d61c044b1678 // indirect
 	golang.org/x/text v0.3.7 // indirect
 	gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c // indirect
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
```
Comment (unknwon) [inline go.mod:15]: This is a direct dependency, please do a `go mod tidy`.

File: binding_test.go [hunk 8]
```diff
@@ -346,6 +346,93 @@ func TestJSON(t *testing.T) {
...
+				payload:    `Username: alice`,
+				handler:    normalHandler,
+				statusCode: http.StatusBadRequest,
+				want:       `Key: "yaml.Password" Error: Field validation for "Password" failed on the "required" tag`,
+			},
+			{
+				name: "fast invoker handler",
+				payload: `Username: alice
+							handler:    fastInvokerHandler,
+			Password: supersecurepassword`,
```
Comment (unknwon) [inline binding_test.go:401]: The formatting here looks weird, why is it looking like this?

**Review by unknwon — CHANGES_REQUESTED** [review]

**Comment by Germiniku** [comment]
Sorry，I made some stupid mistakes.

**Comment by unknwon** [comment]
> Sorry，I made some stupid mistakes.

No need to be sorry at all :)

**Comment by Germiniku** [comment]
I have now increased YAML test coverage to 95.2%

File: binding.go (hunk 1 above)
Comment (unknwon) [reply binding.go:17]: @Germiniku This has not been addressed.

File: binding_test.go (hunk 5 above)
Comment (unknwon) [reply binding_test.go:362]: @Germiniku I think here we are still missing the test case without "custom error handler".

File: binding_test.go [hunk 9]
```diff
@@ -346,6 +346,459 @@ func TestJSON(t *testing.T) {
...
+				}
+				f.Post("/", YAML(yaml{}, opts), func(c flamego.Context) {
+					_, _ = c.ResponseWriter().Write([]byte("Hello world"))
+				})
+
+				resp := httptest.NewRecorder()
+				req, err := http.NewRequest(http.MethodPost, "/", bytes.NewBufferString(test.payload))
+				assert.Nil(t, err)
+
+				req.Header.Set("Content-Type", "application/x-yaml")
```
Comment (unknwon) [inline binding_test.go:434]: Is there a MDN page recommending setting this value? Seems not necessary be put here.

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the followup!

Left few more comments.

**Comment by Germiniku** [comment]
Content-Type application/x-yaml  i have not found on MDN. this is a found on stackoverflow link [https://stackoverflow.com/questions/332129/yaml-media-type](url)

File: binding_test.go [hunk 10]
```diff
@@ -346,6 +346,481 @@ func TestJSON(t *testing.T) {
...
+				resp := httptest.NewRecorder()
+				req, err := http.NewRequest(http.MethodPost, "/", bytes.NewBufferString(test.payload))
+				assert.Nil(t, err)
+				f.ServeHTTP(resp, req)
+				assert.Equal(t, test.statusCode, resp.Code)
+				assert.Equal(t, test.want, resp.Body.String())
+			})
+		}
+	})
+	type address struct {
```
Comment (unknwon) [inline binding_test.go:466]:
```suggestion
	})

	type address struct {
```

File: binding_test.go [hunk 11]
```diff
@@ -346,6 +346,481 @@ func TestJSON(t *testing.T) {
...
+				}
+				f.Post("/", YAML(yaml{}, opts), func(c flamego.Context) {
+					_, _ = c.ResponseWriter().Write([]byte("Hello world"))
+				})
+
+				resp := httptest.NewRecorder()
+				req, err := http.NewRequest(http.MethodPost, "/", bytes.NewBufferString(test.payload))
+				assert.Nil(t, err)
+				f.ServeHTTP(resp, req)
+				assert.Equal(t, test.statusCode, resp.Code)
```
Comment (unknwon) [inline binding_test.go:461]:
```suggestion
				assert.Nil(t, err)

				f.ServeHTTP(resp, req)

				assert.Equal(t, test.statusCode, resp.Code)
```

File: binding_test.go [hunk 12]
```diff
@@ -346,6 +346,481 @@ func TestJSON(t *testing.T) {
...
+			f.Post("/", YAML(user{}), func(form user, errs Errors) {
+				gotForm = form
+				gotErrs = errs
+			})
+
+			resp := httptest.NewRecorder()
+			req, err := http.NewRequest(http.MethodPost, "/", bytes.NewBufferString(test.body))
+			assert.Nil(t, err)
+			f.ServeHTTP(resp, req)
+			test.assertErrors(t, gotErrs)
```
Comment (unknwon) [inline binding_test.go:818]:
```suggestion
			assert.Nil(t, err)

			f.ServeHTTP(resp, req)

			test.assertErrors(t, gotErrs)
```

**Review by unknwon — APPROVED** [review]
Thank you, all looking good!

**Comment by unknwon** [comment]
Congratulations 🎉 and https://github.com/flamego/binding/releases/tag/v1.1.0 has been tagged for this merge!

**Pull request merged**